src/js/select2/data/ajax.js

```diff
@@ -66,6 +66,9 @@ define([
 
     function request () {
       var $request = options.transport(options, function (data) {
+        if (data.results) {
+          data.results = data.results.map(AjaxAdapter.prototype._normalizeItem);
+        }
         var results = self.processResults(data, params);
```

Comment:
```suggestion
        var results = self.processResults(data, params);
        if (results) {
          results = results.map(AjaxAdapter.prototype._normalizeItem);
        }
```

I'm thinking this should happen after the results processing, since in many cases people bring the results into compliance through that method already (as well as doing things like mapping `id` from other fields).

So this would actually get moved down ahead of the callback and would also need to be guarded by `Array.isArray()` based on the debug catch which we have below this block. Sometimes people don't return an array properly so we have to kind of just roll with it and let it happen.

Reply:
Thanks, ok - that makes sense :+1:

Reply:
(`Array.isArray` guard added, and a single conditional clause is now used to check the post-processed results in the correct-data path, and for debug error reporting in the malformed-data path)